pkg/datasources/v1/registry.go

```diff
@@ -0,0 +1,66 @@
+// SPDX-FileCopyrightText: Copyright 2024 The Minder Authors
+// SPDX-License-Identifier: Apache-2.0
+
+package v1
+
+import (
+	"errors"
+	"fmt"
+
+	"github.com/puzpuzpuz/xsync/v3"
+)
+
+var (
+	// ErrDuplicateDataSourceFuncKey is the error returned when a data source
+	// function key is already registered.
+	ErrDuplicateDataSourceFuncKey = errors.New("duplicate data source function key")
+)
+
+// DataSourceRegistry is the interface that a data source registry must implement.
+// It provides methods to register a data source and get all functions that
+// data sources provide globally.
+type DataSourceRegistry struct {
+	r *xsync.MapOf[DataSourceFuncKey, DataSourceFuncDef]
+}
+
+// NewDataSourceRegistry creates a new data source registry.
+func NewDataSourceRegistry() *DataSourceRegistry {
+	return &DataSourceRegistry{
+		r: xsync.NewMapOf[DataSourceFuncKey, DataSourceFuncDef](),
+	}
+}
+
+// RegisterDataSource registers a data source with the registry.
+// Note that the name of the data source must be unique.
+func (reg *DataSourceRegistry) RegisterDataSource(name string, ds DataSource) (err error) {
+	for key, f := range ds.GetFuncs() {
+		funckey := makeKey(name, key)
+		if _, ok := reg.r.Load(funckey); ok {
+			return fmt.Errorf("%w: %s", ErrDuplicateDataSourceFuncKey, funckey)
+		}
+
+		// We only flush the store if there was no error
+		defer func() {
+			if err == nil {
+				reg.r.Store(funckey, f)
+			}
+		}()
+	}
+
+	return nil
+}
+
+// GetFuncs returns all functions that data sources provide globally.
```

Comment:
This could eventually pull from the database to get project-specific data sources, right?

Reply:
That's right, the idea is that we'd do an initial parsing from the db, turn this into in-memory objects, and then pass these around.